epochStart/metachain/systemSCs.go

```diff
@@ -343,6 +354,48 @@ func (s *systemSCProcessor) unStakeOneNode(blsKey []byte, epoch uint32) error {
 	return nil
 }
 
+func (s *systemSCProcessor) updateDelegationContracts(mapOwnerKeys map[string][][]byte) error {
+	sortedDelegationsSCs := make([]string, 0, len(mapOwnerKeys))
+	for address := range mapOwnerKeys {
+		if s.shardCoordinator.ComputeId([]byte(address)) != core.MetachainShardId {
+			continue
+		}
+		sortedDelegationsSCs = append(sortedDelegationsSCs, address)
+	}
+
+	sort.Slice(sortedDelegationsSCs, func(i, j int) bool {
+		return sortedDelegationsSCs[i] < sortedDelegationsSCs[j]
+	})
+
+	for _, address := range sortedDelegationsSCs {
+		vmInput := &vmcommon.ContractCallInput{
+			VMInput: vmcommon.VMInput{
+				CallerAddr: s.endOfEpochCallerAddress,
+				Arguments:  mapOwnerKeys[address],
+				CallValue:  big.NewInt(0),
+			},
+			RecipientAddr: []byte(address),
+			Function:      "unStakeAtEndOfEpoch",
+		}
+
+		vmOutput, err := s.systemVM.RunSmartContractCall(vmInput)
+		if err != nil {
+			return err
+		}
+		if vmOutput.ReturnCode != vmcommon.Ok {
```

Comment:
this process will stop if a delegator does not have enough funds to re-delegate because he/she received too few rewards from the past epoch.

Reply:
this is unStakeAtEndOfEpoch. no redelegation.

Reply:
this will not error.

Reply:
right